Continuous performance unit testing

gunnarmorling opened the pull request.

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
```diff
@@ -0,0 +1,424 @@
+
+As we'll see in a bit, JfrUnit integrates nicely with the Java streams API,
+simplifying the filtering and aggregation of recorded events before matching them against asserted values.
+It also persists the JFR recording of each test method,
```
"It also persists" -> does refer to what exactly here? I assume it == JfrUnit itself but that is a bit unclear IMHO :-)

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
to keep an eye **on**

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
... *and* in turn
throughput

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
nice to learn about this while reading 👍

gunnarmorling replied:
Ha, thanks. Was also a nice learning while writing :)

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
... on a test ~that'll~ to spot ...

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
something seems wrong with the method params here? int given but Random expected?

gunnarmorling replied:
Wow, good spot!

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
see comment above -> I assume the code was changed but not this snippet ?

gunnarmorling replied:
Yes, it evolved :) Should be fixed with my next push.

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
I would find one call to .filter(...) and combined expression more natural here? WDYT?

gunnarmorling replied:
Hum, hum. Couldn't use the method refs then, I think?

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
I'm wondering if it would make sense to define a lower threshold as well since in the text for <3> you talk about being *within a boundary*, so +/- ~1% ? Of course, you could argue that doing less allocations is "always" fine ...

gunnarmorling replied:
That's a great point actually, touching on something that was on my mind. Say allocation improves (new JDK, etc.) and then gets worse again, e.g. from 100 -> 60 -> 80. Right now, we wouldn't spot that regression from 60 to 80, due to that initial improvement. I'll add a note on that. Thanks!

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
maybe change the overly positive associated "wow" to something else?
also, "more" is used 2x here which reads a bit weird to me

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
would it make sense and be possible to exclude these internal events from the recording? if not during, maybe as a post-processing step or is it simply not worth it?

gunnarmorling replied:
It cannot be excluded, as it's required as sort of a synchronization barrier for making sure all previously produced events have been received (this all is async). I suppose you could post-process the recording, but I don't think it's worth it actually.

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
... <ins>in</ins> its own right ...

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
throughput
... **in** (?) a production like environment.

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
...the call *against* (?) the User REST API...

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
... for an *alternate* (?) REST client.

gunnarmorling replied:
I think alternative actually is correct here. alternate as adverb means "wechselnd".

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
note: I stopped processing here ;-)

hpgrahsl commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
todo: replace placeholder event type/name ;-)

hpgrahsl reviewed (commented):
really enjoyed this early stage read! my comments are mostly referring to trivial things like typos etc. plus a few questions. nothing serious I would say.

gunnarmorling commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
```diff
@@ -0,0 +1,424 @@
+E.g. assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
+When run in a virtualized or containerized CI environments, such tests are are prone to failures due to concurrent load of other applications and jobs.
+
+This post introduces https://github.com/gunnarmorling/jfrunit[JfrUnit], which offers a fresh angle to this topic by supporting assertions not on metrics like latency/throughput themselves, but on indirect metrics which may impact those. Based on https://openjdk.java.net/jeps/328[JDK Flight Recorder] events, JfrUnit allows you define and execute assertions e.g. against expected memory allocation,  ...
```
Good question. You could have N+1 also without ORM, but you're right that this is the most common cause. Will reword.

gunnarmorling commented on content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc:
```diff
@@ -0,0 +1,424 @@
+
+This post is the first one in a series of three:
+
+* Part one provides an overview over continuous performance unit testing with JfrUnit,
```
Yeah, I'm still unsure. It's *JUnit* of course, but not unit testing strictly speaking, I suppose. Will need to think about it.

gunnarmorling commented:
Hey @hpgrahsl, thanks a lot for that first round of review! I've addressed most of your remarks, and I've also added the missing parts to the post. I.e. it's ready now for a review of that remainder, should you have the time and interest. There's also a [rendered preview](https://gunnarmorling-morling-dev-preview-pr-15.surge.sh/blog/towards-continuous-performance-regression-testing/) available now. I'm planning to do some more polishing and also updates to the images, but overall, it's 95% of what I had in mind. Thank you so much!

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+
+Functional unit and integration tests are a standard tool of any software development organization,
+helping not only to ensure correctness of newly implemented code,
+but also to identify regressions -- bugs to existing functionality introduced by a code change.
```
```suggestion
but also to identify regressions -- bugs in existing functionality introduced by a code change.
```

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+but also to identify regressions -- bugs to existing functionality introduced by a code change.
+The situation looks different though when it comes to regressions related to non-functional requirements, in particular performance-related ones:
+How to detect increased response times in a web application?
+How to identify decreased throughput?
```
While I understand what you mean, I'd say that these specific examples are a bit off since they are very much metrics which I would NOT test using your new tool?

There's a need for testing at both levels:
 - very micro, such as "This sort operation here can be performed in less than Z memory allocated even for an N sized array"
 - system wide impact, such as system design and integration is such that overall throughput (on certain machine) is within X,Y with some margins.

I'd use a tool like JfrUnit for the first cathefory only, it seems a slippery slope to try abusing it for beyond this and that's probably a claim I'd be umcofortable with :)

gunnarmorling replied:
> There's a need for testing at both levels

Agreed. I think JfrUnit's testing approach can play a role for both, though. For the second, it wouldn't help you answering that question "overall throughput is within X,Y with some margins" directly, but it would help you to identify potential regressions, going against that goal. It should be more clear in the discussion towards the end, perhaps I need to reword here a bit, too.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+How to identify decreased throughput?
+
+These aspects are typically hard to test in an automated and reliable way in the development workflow,
+as they are dependent on the underlying hardware and the workload of an application.
```
But best to split what you're going to measure here in cathegories. Since JfrUnit is about "indirect" metrics, you can't really verify that not having a certain `Sleep(1000)` or not having a certain GC event will necessarily give you the throughput you're after.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+
+These aspects are typically hard to test in an automated and reliable way in the development workflow,
+as they are dependent on the underlying hardware and the workload of an application.
+For instance assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
```
Maybe qualify this? People who do care a bit more about performance might not be doing their homework on a laptop, and might be running Linux on their daily driver for this very reason of actually emulating a production environment closely.

Sure ideally one wants to test on the same "production hardware" where ultimately the code is going to run, but when working on flexible libraries which are going to be deployed in a variety of systems, that's as good as it can get.

gunnarmorling replied:
Reworded slightly.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+These aspects are typically hard to test in an automated and reliable way in the development workflow,
+as they are dependent on the underlying hardware and the workload of an application.
+For instance assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
+When run in a virtualized or containerized CI environment, such tests are prone to failures due to concurrent load of other applications and jobs.
```
I'm assuming you are suggesting the idea of running performance tests on virtualized / shared resources here, for example we have some tests which send an alert if some metric gets beyond some % trheshold of difference with previous averages. Might want to clarify that?

In which case I'd say s/failures/ severe measurement distortions / ?

There's also other challenges: valid bugfixes and good improvements which tend to slowly get you closer to thethreshold of failure will regularly erode safety margins, eventually triggering CI failures for totally innocent changes; for example I've seen a typo in a README trigger a performance failure - because such metrics need constant "gardening" to be useful.

gunnarmorling replied:
> I'd say s/failures/ severe measurement distortions / ?

+1 I like that.

> There's also other challenges [...]

Yes, great point. Added this to the section at the end about possible future additions. Also discussing with John O'Hara about this who has similar ideas.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+For instance assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
+When run in a virtualized or containerized CI environment, such tests are prone to failures due to concurrent load of other applications and jobs.
+
+This post introduces https://github.com/gunnarmorling/jfrunit[JfrUnit], which offers a fresh angle to this topic by supporting assertions not on metrics like latency/throughput themselves, but on _indirect metrics_ which may impact those.
```
+100 awesome

But could use an example of "indirect metrics" and why they matter?

gunnarmorling replied:
It's clarified in the next sentence. Also reworded it a bit to make the point stronger.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+
+This post introduces https://github.com/gunnarmorling/jfrunit[JfrUnit], which offers a fresh angle to this topic by supporting assertions not on metrics like latency/throughput themselves, but on _indirect metrics_ which may impact those.
+Based on https://openjdk.java.net/jeps/328[JDK Flight Recorder] events, JfrUnit allows you define and execute assertions e.g. against expected memory allocation, database I/O, or number of executed SQL statements, for a given workload.
+Starting off from a defined base line, future failures of such assertions are indicators for potential performance regressions in an application, as a code change may have introduced higher GC pressure,
```
I don't think there's such a thing as "GC pressure". you probably mean excessive allocations?

What I'd see your tool most useful for is to introduce regression tests for specific optimisations which have already been identified in the past as problematic under certain workloads. For example, as you well know, there's some code areas of Hibernate ORM which have been carefully tuned for being very lean on allocations, while in 90% of the codebase this doesn't matter (or its importance has not been identified yet).

When one implements such an optimisation, it requires a specialized benchmark to verify that the fix has been performed, but we don't have the means to protect the feature against future regressions.

In such a scenario I'd think JfrUnit would be "gold", as one has a deep understanding of the specific metrics which matter in that case, and can encode the goal as "indirect metrics" to be tested more easily in the future.

Of course that doesn't guarantee any figure to be delivered on the same benchmark again, but it's much better than not doing anything at all.

gunnarmorling replied:
Yes, that's exactly the use case I have in mind. I think you would benefit a lot from JfrUnit in Hibernate by using it to set up a performance regression test suite. But beyond that, I think it also can be interesting for application developers, which is what I'm focusing a bit more in this post (solely due to the fact that that this is a much larger audience).

gunnarmorling replied:
> I don't think there's such a thing as "GC pressure". you probably mean excessive allocations?

The latter is leading to the former :) Seeing quite a few usages of that term, including JDK docs, so I think it's legit.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+This post introduces https://github.com/gunnarmorling/jfrunit[JfrUnit], which offers a fresh angle to this topic by supporting assertions not on metrics like latency/throughput themselves, but on _indirect metrics_ which may impact those.
+Based on https://openjdk.java.net/jeps/328[JDK Flight Recorder] events, JfrUnit allows you define and execute assertions e.g. against expected memory allocation, database I/O, or number of executed SQL statements, for a given workload.
+Starting off from a defined base line, future failures of such assertions are indicators for potential performance regressions in an application, as a code change may have introduced higher GC pressure,
+the retrieval of unneccessary data from the database, or SQL problems commonly induced by ORM tools, like N+1 SELECT statements.
```
well N+1 is meh.. it's certainly good to test against it (and maybe we should see to make this easier with some helper and a blog) but that could be done in much simpler / traditional ways.

I feel it's a bit distracting from the real potential of the Jfr metrics.

gunnarmorling replied:
Happy to discuss another example around Hibernate/SQL, if you have one? N+1 was the first thing that came to my mind. It's going to be discussed in a follow-up post (in January), so you got some time to think about it ;)

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+
+    ExpectedEvent = event("jdk.ThreadSleep").
+        with("time", Duration.ofSeconds(1)));
+    assertThat(jfrEvents).contains(event); 
```
is this precise enough to reliably work? I guess you should expand a little on how the event attributes are matched.

gunnarmorling replied:
Added an explantory sentence.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+    
+    ExpectedEvent = event("jdk.GarbageCollection") // <4>
+        .with("cause", "System.gc()"));
+    assertThat(jfrEvents).contains(event); 
```
pretty cool. So the `with("cause"` created a restriction the matcher ? I mean, I take it that this wouldn't fail if there was a naturally occuring GC event accidentally being recorded as well?

gunnarmorling replied:
Yes, exactly. `with()` sets up the given event attribute to be matched against. If there'd be another GC with another cause, this wouldn't cause this assertion to fail. We might think about adding `containsNot()` though, this could be useful actually. Captured in https://github.com/gunnarmorling/jfrunit/issues/8.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```diff
@@ -0,0 +1,594 @@
+----
+<1> `@JfrEventTest` marks this as a JfrUnit test, activating its extension
+<2> All JFR event types to be recorded must be enabled via `@EnableEvent`
+<3> After running the test logic, `awaitEvents()` must be invoked as a synchronization barrier,
```
so, design question.. why not expressing all expectations declaratively rather than have the user deal with the need for barriers?

gunnarmorling replied:
Yes, I was (am) considering that, as it makes the barrier issue go away, as you say. But declarative makes for much less flexibility in terms of expressing those expected values, you'd have to come up with a solution for summing up, aggregations, custom events etc. You'll land in "programming in annotations" land pretty quickly. So I'm leaning towards the imperative approach.

Sanne replied:
ah yes I understand it better after having seen the second example.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```suggestion
Since this space is exclusively owned by the thread, new objects within a TLAB can happen without costly synchronization with other threads.
```

gunnarmorling replied:
Nice.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```suggestion
In addition, large objects will likely need to be directly allocated outside of the more efficient TLAB space.
```

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
hum I don't remember the limit defaults, but I think it's a bit misleading to let people think it will just keep growing, as the whole point of our most important optimisations is to not exceed the limit.

gunnarmorling replied:
I'm not quite following on that one; what is the "it" in "it will just keep growing"? The TLAB doesn't grow, it will be used up by allocations, and when the next allocation doesn't fit into it, the thread will get a new TLAB for the next set of allocations. You won't avoid that, unless you do zero allocations or allocate in your own self-managed `byte[]`.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
Are you sure? I do remember chasing TLAB allocation in more detail, but indeed this was in the older JFR and I never really understood why this metric was now. Maybe that was a mistake and it was removed later?

Sanne replied:
cc/ @stalep

gunnarmorling replied:
Yes, pretty sure. The "inside TLAB" event gives you the new TLAB size and the size of the object which caused this (hence the notion of sampling). The "outside TLAB" gives you the object size, perhaps that's what you have in mind?

Sanne replied:
ah yes, I simply misunderstood what you meant. _sampling_ actually is an option. I misread this as suggesting one couldn't get any clue about what was getting into TLAB.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
it's a bit weird that the API seems completely different here? Rather than "selecting" by key now I get to stream all events?

Sanne replied:
(I understand you need it here to do the accumulation, but I'm suggesting it needs a warning that you'll be using a different approach)

gunnarmorling replied:
Good point. Clarified.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
hugh, this really need to allow zooming in

gunnarmorling replied:
Yes, working on making the images click-zoomable.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
:100:

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```suggestion
It's also worth examining the application's garbage collection behavior.
```

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
I think I'd like also:

1. implement your own custom JFR events
2. use this to test your custom events

:-)

gunnarmorling replied:
Done.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```suggestion

```

I don't think it's that bad. Ideally, make it a lazy blob...

gunnarmorling replied:
Yes, mentioning that as an option later on.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
add : "although the Hibernate team wants to look into this " .. ;)

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```suggestion
Regressions in such impacting metrics, e.g. the amount of allocated memory, or bytes read from a database,
```

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
```suggestion
* *Pro-active identification of performance issues:* Asserting a metric like memory allocation can help to identify future performance problems before they actual materialize; while the additional allocation rate may make no difference with the system's load as of today, it may negatively impact latency and throughput as the system reaches its limits with increased load; being able to identify the increased allocation rate early on allows for a more efficient handling of the situation while working on the code, compared to when finding out about such regression only later on
```

Sanne replied:
Also might want to consider: in a complete production system there might be more components being integrated then the ones we have in our testbeds.

When measuring allocation rate of components individually, they might not seem problematic as it's easier to keep the allocation budget within the cheaper TLAB range, so while their individual allocation rate might not seem to have a problematic impact on throughput measurements performed on simple benchmarks, an excessive use of allocations could still translate in problematic bottlenecks on a more complex system as the same TLAB region is shared with other libraries.

gunnarmorling replied:
That's an interesting one; any idea how you'd identify such bottleneck? I don't think it needs discussion in this post (it's super long already), but would like to better understand and see what perhaps even can be done in JfrUnit itself towards that end.

Sanne commented on content/blog/towards-continuous-performance-regression-testing.asciidoc:
nice - but who is he? A link to him, twitter maybe?

gunnarmorling replied:
Yes, I'll add links to him, you, and other reviewers, unless you oppose to being mentioned :)

gunnarmorling commented:
Ok, so I think I've addressed the critical review remarks by all of you. Thank you so much. Latest update just pushed, going to look into zoomable images next.

gunnarmorling commented:
Some more word-smithing and fixing. Pushing now. Thank you all, I'm deeply grateful for all the feedback you provided on such short notice 🙏 !
